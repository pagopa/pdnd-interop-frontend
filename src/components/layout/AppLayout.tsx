import React from 'react'
import { Box, Stack } from '@mui/material'
import type { SxProps } from '@mui/material'
import { SideNav } from '@/components/layout'
import { Breadcrumbs } from './Breadcrumbs'

type AppLayoutProps = {
  children: React.ReactNode
  hideSideNav?: boolean
  sx?: SxProps
}

export const AppLayout: React.FC<AppLayoutProps> = ({ children, hideSideNav, sx }) => {
  if (hideSideNav) {
    return (
      <Box
        component="main"
        sx={{
          px: 3,
          height: '100%',
          py: 2,
          ...sx,
        }}
        bgcolor="#FAFAFA"
      >
        <Box sx={{ maxWidth: 1280, mx: 'auto' }}>{children}</Box>
      </Box>
    )
  }

  return (
    <Box sx={{ flexGrow: 1, height: '100%' }}>
      <Stack direction="row" sx={{ height: '100%', overflowX: 'hidden' }}>
        <SideNav />
        <Box
          sx={{
            px: 3,
            py: 2,
            flexGrow: 1,
            position: 'relative',
            '::after': {
              content: '""',
              position: 'absolute',
              right: 0,
              top: 0,
              bgcolor: 'background.default',
              width: 10000,
              height: '100%',
              transform: 'translate(100%, 0)',
            },
          }}
          bgcolor="#FAFAFA"
        >
          <Box component="main" sx={{ height: '100%', maxWidth: 1280 }}>
            <Breadcrumbs />

            {children}
          </Box>
        </Box>
      </Stack>
    </Box>
  )
}
