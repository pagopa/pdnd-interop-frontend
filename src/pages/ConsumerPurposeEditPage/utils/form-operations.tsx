import {
  CheckboxGroup,
  RadioGroup,
  Select,
  TextField,
} from '@/components/shared/ReactHookFormInputs'
import type { InputOption } from '@/types/common.types'
import React from 'react'
import identity from 'lodash/identity'
import type {
  Dependency,
  DynamicFormOperations,
  Questions,
  QuestionV1,
  QuestionV2,
} from '../types/risk-analysis.types'
import { RiskAnalysisSwitch } from '../components/PurposeEditStep2RiskAnalysis/RiskAnalysisSwitch'

export const dynamicFormOperationsVersions: DynamicFormOperations = {
  '1.0': {
    getUpdatedQuestions: (values, riskAnalysis) => {
      const updatedQuestions = riskAnalysis.questions.reduce((acc, next) => {
        // We are sure we can only queue the questions that come after the current one
        // because the dependencies are always on the questions that come before.
        // E.g. question 3 can only have question 1 and/or 2 as a dependency,
        // not question 4 and subsequent
        const queuedQuestions = Object.values(acc).map((d) => d.id)

        // Check the dependencies of this questions
        const satisfiesAllDependencies = next.dependencies.every((d) => {
          // If the question is a dependency
          const isInArray = queuedQuestions.includes(d.id)
          const id = d.id as string

          // and if the field currently has the same value as stated in the dependency
          const hasSameValue = Array.isArray(values[id])
            ? (values[id] as Array<unknown>).includes(d.value)
            : values[id] === d.value

          return isInArray && hasSameValue
        })

        // Then it needs to be added as a new question to ask
        if (satisfiesAllDependencies) {
          acc[next.id] = next
        }

        return acc
      }, {} as Questions)
      return updatedQuestions
    },

    buildForm: (questions, _, lang, t) => {
      const questionKeys = Object.keys(questions)

      const formComponents = questionKeys.map((id, i) => {
        const { type, label, options, infoLabel } = questions[id] as QuestionV1

        const sx = questionKeys.length - 1 === i ? { mb: 0 } : {}
        const inputOptions = options ? options.map((o) => ({ ...o, label: o.label[lang] })) : []
        const commonProps = {
          key: id,
          name: id,
          label: label[lang],
          infoLabel: infoLabel ? infoLabel[lang] : undefined,
          sx,
        }

        switch (type) {
          case 'text':
            return <TextField {...commonProps} rules={{ required: true }} />

          case 'select-one':
            return (
              <Select
                {...commonProps}
                options={inputOptions}
                emptyLabel={t('edit.step2.emptyLabel')}
                rules={{ required: true }}
              />
            )

          case 'checkbox':
            const errorMessage =
              options && options?.length > 1
                ? t('edit.step2.multiCheckboxField.validation.mixed.required')
                : t('edit.step2.singleCheckboxField.validation.mixed.required')
            return (
              <CheckboxGroup
                {...commonProps}
                options={inputOptions}
                rules={{
                  validate: (value) =>
                    (typeof value !== 'undefined' && value.length > 0) || errorMessage,
                }}
              />
            )

          case 'radio':
            return <RadioGroup {...commonProps} options={inputOptions} rules={{ required: true }} />
        }
      })

      return { formComponents: formComponents, isSubmitBtnDisabled: false }
    },
  },
  '2.0': {
    getUpdatedQuestions: (values, riskAnalysis) => {
      const questions = riskAnalysis.questions as Array<QuestionV2>

      // Filters all the questions that not satisfies the dependency inside the "dependencies" property
      const updatedQuestions = questions.filter(({ dependencies }) => {
        function satisfiesDependency(dep: QuestionV2['dependencies'][0]) {
          if (Array.isArray(values[dep.id])) {
            return (values[dep.id] as Array<unknown>).includes(dep.value)
          }
          if (Array.isArray(dep.value)) {
            return dep.value.includes(values[dep.id])
          }
          return values[dep.id] === dep.value
        }
        // If length of dependencies array is 0 the "every" method returns true
        return dependencies.every(satisfiesDependency)
      })

      // Array<Question> -> Record<"id", Question>
      return updatedQuestions.reduce((acc, next) => ({ ...acc, [next.id]: next }), {} as Questions)
    },

    buildForm: (questions, formMethods, lang, t) => {
      const values = formMethods.getValues()

      const formComponents: Array<React.ReactNode> = []

      const questionIds = Object.keys(questions)

      type QuestionOption = Exclude<QuestionV2['options'], undefined>[0]

      function checkOptionDependency(dependency: Dependency) {
        const currentDepValue = values[dependency.id]

        const hasDependency =
          currentDepValue === dependency.value ||
          (Array.isArray(currentDepValue) &&
            (currentDepValue as string[]).includes(dependency.value as string))

        return hasDependency
      }

      function parseOption(option: QuestionOption, { hideOption, id, defaultValue }: QuestionV2) {
        // if the key "hideOption" is present in the question object and the conditions are satisfied
        // the option will be not added to the array of options
        const shouldHideOption =
          hideOption &&
          hideOption[option.value] &&
          hideOption[option.value].some(checkOptionDependency)

        if (!shouldHideOption) {
          return { value: option.value, label: option.label[lang] }
        }

        if ((values[id] as string[])?.includes(option.value)) {
          formMethods.setValue(id, defaultValue)
        }
      }

      function buildFormQuestionComponents(
        { id, type, ...question }: QuestionV2,
        inputOptions: Array<InputOption>,
        isLast: boolean
      ) {
        const questionComponents: Array<React.ReactNode> = []

        const maxLength = question?.validation?.maxLength
        const isRequired = question.required
        const isMultipleChoice = question.dataType === 'multi'

        let label = question.label[lang]
        let infoLabel = question.infoLabel && question.infoLabel[lang]

        if (maxLength) {
          const maxLengthLabel = t('edit.step2.validation.maxLength', { num: maxLength })
          if (infoLabel) {
            infoLabel += `. ${maxLengthLabel}`
          } else {
            infoLabel = maxLengthLabel
          }
        }

        const labelValidation: Array<string> = []

        if (isRequired) {
          labelValidation.push(t('edit.step2.validation.required'))
        }

        if (isMultipleChoice) {
          labelValidation.push(t('edit.step2.validation.multipleChoice'))
        }

        if (labelValidation.length > 0) {
          label += ` (${labelValidation.join(', ')})`
        }

        const sx = isLast ? { mb: 0 } : {}
        const commonProps = { key: id, name: id, label, infoLabel, sx }

        switch (type) {
          case 'text':
            questionComponents.push(
              <TextField {...commonProps} inputProps={{ maxLength }} rules={{ required: true }} />
            )
            break
          case 'select-one':
            questionComponents.push(
              <Select
                {...commonProps}
                options={inputOptions}
                emptyLabel={t('edit.step2.emptyLabel')}
                rules={{ required: true }}
              />
            )
            break
          case 'checkbox':
            questionComponents.push(
              <CheckboxGroup
                {...commonProps}
                options={inputOptions}
                rules={{
                  validate: (value) =>
                    (typeof value !== 'undefined' && value.length > 0) ||
                    t('edit.step2.multiCheckboxField.validation.mixed.required'),
                }}
              />
            )
            break
          case 'radio':
            questionComponents.push(
              <RadioGroup {...commonProps} options={inputOptions} rules={{ required: true }} />
            )
            break
          case 'switch':
            questionComponents.push(
              <RiskAnalysisSwitch
                {...commonProps}
                options={inputOptions}
                rules={{
                  validate: (value) =>
                    (typeof value === 'boolean' && value === true) ||
                    t('edit.step2.riskAnalysisSwitch.validation.boolean.isValue'),
                }}
              />
            )
            break
        }

        return questionComponents
      }

      for (let i = 0; i < questionIds.length; i++) {
        const questionId = questionIds[i]

        const question = questions[questionId] as QuestionV2

        const inputOptions = (question.options
          ?.map((option) => parseOption(option, question))
          .filter(identity) || []) as Array<InputOption>

        formComponents.push(
          ...buildFormQuestionComponents(question, inputOptions, questionIds.length - 1 === i)
        )
      }

      return { formComponents, isSubmitBtnDisabled: false }
    },
  },
}

export function getFormOperations(version: string) {
  const dynamicFormOperations = dynamicFormOperationsVersions[version]

  if (!dynamicFormOperations) {
    throw new Error('There is no business logic for this specific risk analysis data form')
  }

  return dynamicFormOperations
}
