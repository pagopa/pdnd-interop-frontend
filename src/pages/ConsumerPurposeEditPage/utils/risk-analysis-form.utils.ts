import type { Dependency, FormConfigQuestion, HideOption } from '@/api/api.generatedTypes'
import type { AnswerValue, Answers, Questions } from '../types/risk-analysis-form.types'

/**
 * Transform the answers from the backend to the frontend format
 * The answer from the backend arrives as an array of strings, but the frontend
 * needs a specific format depending on the `visualType` value.
 *
 * - `switch` needs a boolean
 * - `checkbox` needs an array of strings
 * - the rest needs a string
 *
 * @param backendAnswer The answer from the backend
 * @param question The question that the answer belongs to
 * @returns The answer in the type that the frontend needs
 */
export function getFrontendAnswerValue(
  backendAnswer: Array<string>,
  visualType?: FormConfigQuestion['visualType']
): AnswerValue {
  if (visualType === 'switch') {
    return backendAnswer[0] === 'true'
  }

  if (visualType === 'checkbox') {
    return backendAnswer
  }

  return backendAnswer[0]
}

/**
 * Transform the answers from the frontend to the backend format
 * The backend always needs an array of strings.
 *
 * @param answer The answer from the frontend
 * @returns The answer in the type that the backend needs
 * */
export function getBackendAnswerValue(answer: AnswerValue): Array<string> {
  if (typeof answer === 'boolean') {
    return [answer.toString()]
  }

  if (Array.isArray(answer)) {
    return answer
  }

  if (typeof answer === 'string') {
    return [answer]
  }

  throw new Error(`Invalid answer type: ${typeof answer} \n\n ${JSON.stringify(answer, null, 2)}`)
}

/**
 * Returns the valid answers that should be sent to the backend.
 * The answers that should be sent to the backend are only the answers
 * to the questions that are currently visible to the user.
 * The answers to the questions that are not visible to the user are not sent.
 *
 * @param currentQuestionsIds - the ids of the questions that are currently visible to the user
 * @param answers - the answers to all the questions
 * @returns the valid answers that should be sent to the backend
 * */
export function getValidAnswers(currentQuestionsIds: Array<string>, answers: Answers) {
  return Object.keys(answers).reduce<Record<string, Array<string>>>((acc, key) => {
    // If the question is actually visible, add it to the valid answers
    if (currentQuestionsIds.includes(key)) {
      return { ...acc, [key]: getBackendAnswerValue(answers[key]) }
    }
    return acc
  }, {})
}

/**
 * Returns true if the dependency is satisfied, false otherwise.
 *
 * @param dependency - the dependency
 * @param answers - the actual form values
 * @returns `true` if the dependency is satisfied, `false` otherwise
 * */
export function isDependencySatisfied(dependency: Dependency | HideOption, answers: Answers) {
  const answer = answers[dependency.id]
  if (Array.isArray(answer)) {
    return answer.includes(dependency.value)
  }

  return answer === dependency.value
}

/**
 * Returns the updated question data.
 *
 * @param answers - the actual form values
 * @param riskAnalysis - the risk analysis document
 *
 * @returns the updated questions data
 * */
export function getUpdatedQuestions(
  answers: Answers,
  riskAnalysisQuestions: FormConfigQuestion[]
): Questions {
  // Filters all the questions that not satisfies the dependency inside the "dependencies" property
  return riskAnalysisQuestions.reduce<Questions>((acc, question) => {
    const doesSatisfyDeps = question.dependencies.every((dependency) =>
      isDependencySatisfied(dependency, answers)
    )
    if (doesSatisfyDeps) {
      acc[question.id] = question
    }
    return acc
  }, {})
}

/**
 * Returns the default values for the risk analysis form.
 * The default values are the values that the form should have when it is first loaded.
 * The default values are the values that the user has already answered, if any.
 * If the user has not answered any question, the values are the default values
 * defined in the risk analysis document.
 *
 * @param riskAnalysisConfig the risk analysis document
 * @param backendAnswers the answers data, taken from the Purpose
 * @returns the default values for the risk analysis form
 */
export function getRiskAnalysisDefaultValues(
  riskAnalysisConfigQuestions: FormConfigQuestion[],
  backendAnswers?: Record<string, Array<string>>
): Answers {
  return riskAnalysisConfigQuestions.reduce<Answers>((acc, question) => {
    const answer = backendAnswers?.[question.id] ?? question.defaultValue
    acc[question.id] = getFrontendAnswerValue(answer, question.visualType)

    if (!acc[question.id] && question.dataType === 'FREETEXT') {
      acc[question.id] = ''
    }

    return acc
  }, {})
}
