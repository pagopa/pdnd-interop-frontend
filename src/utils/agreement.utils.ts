import type { AgreementSummary } from '@/types/agreement.types'
import type { ProviderOrConsumer } from '@/types/common.types'

export const canAgreementBeUpgraded = (agreement: AgreementSummary, mode: ProviderOrConsumer) => {
  if (!agreement || mode !== 'consumer') return false

  return !!(
    agreement.eservice.activeDescriptor &&
    agreement.eservice.activeDescriptor.state === 'PUBLISHED' &&
    agreement.eservice.activeDescriptor.version > agreement.eservice.version &&
    agreement.state !== 'ARCHIVED'
  )
}
