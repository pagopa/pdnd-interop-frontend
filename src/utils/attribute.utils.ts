import type { FrontendAttributes, PartyAttribute } from './../types/attribute.types'
import has from 'lodash/has'
import type { EServiceAttribute, EServiceAttributes } from '@/api/api.generatedTypes'
import identity from 'lodash/identity'

export function checkEServiceAttributesOwnership(
  partyAttributes: Array<PartyAttribute>,
  eserviceAttributes: Array<EServiceAttribute>
) {
  const partyAttributesIds = partyAttributes.filter((p) => p.state === 'ACTIVE').map((p) => p.id)

  const hasAllAttributes = eserviceAttributes.every((eserviceAttribute) => {
    if (has(eserviceAttribute, 'single')) {
      const match = partyAttributesIds.find((id) => id === eserviceAttribute.single?.id)
      return Boolean(match)
    }

    const match = partyAttributesIds.find((id) => {
      const groupIds = eserviceAttribute.group?.map((a) => a.id)
      return groupIds?.includes(id)
    })
    return Boolean(match)
  })

  return hasAllAttributes
}

export function remapEServiceAttributes(backendAttributes: EServiceAttributes): FrontendAttributes {
  const mappedAttributes: FrontendAttributes = Object.keys(backendAttributes).reduce(
    (acc, attributeType) => {
      const mapped = backendAttributes[attributeType as keyof EServiceAttributes]
        .map((attribute: EServiceAttribute) => {
          const isSingle = has(attribute, 'single')

          if (isSingle) {
            const single = attribute.single!
            const { id, explicitAttributeVerification, description, name } = single
            return {
              attributes: [{ id, description, name, kind: attributeType.toUpperCase() }],
              explicitAttributeVerification,
            }
          }

          const isGroup = has(attribute, 'group')

          if (isGroup) {
            const group = attribute.group
            return {
              attributes: [
                ...group!.map(({ id, description, name }) => ({
                  id,
                  description,
                  name,
                  kind: attributeType.toUpperCase(),
                })),
              ],
              explicitAttributeVerification: group![0].explicitAttributeVerification,
            }
          }
        })
        .filter(identity)
      return { ...acc, [attributeType]: mapped }
    },
    { certified: [], verified: [], declared: [] }
  )

  return mappedAttributes
}
