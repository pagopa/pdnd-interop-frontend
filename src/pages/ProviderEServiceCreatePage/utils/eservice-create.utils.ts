import type {
  AgreementApprovalPolicy,
  DescriptorAttributes,
  ProducerEServiceDescriptor,
} from '@/api/api.generatedTypes'
import type { RemappedEServiceAttributes } from '@/types/attribute.types'
import { getKeys } from '@/utils/array.utils'
import { remapEServiceAttributes } from '@/utils/attribute.utils'
import isEqual from 'lodash/isEqual'

type DescriptorDataStep2ToCompare = {
  voucherLifespan: number
  audience: string[]
  agreementApprovalPolicy: AgreementApprovalPolicy
  version: string
  description: string
  dailyCallsPerConsumer: number
  dailyCallsTotal: number
}

export function compareDescriptorsStep2DataIfEquals(
  newDescriptorData: DescriptorDataStep2ToCompare,
  descriptorToCompare: ProducerEServiceDescriptor
) {
  const descriptorDataToCompare: DescriptorDataStep2ToCompare = {
    voucherLifespan: descriptorToCompare.voucherLifespan,
    audience: descriptorToCompare.audience,
    agreementApprovalPolicy: descriptorToCompare.agreementApprovalPolicy,
    version: descriptorToCompare.version,
    description: descriptorToCompare.description ?? '',
    dailyCallsPerConsumer: descriptorToCompare.dailyCallsPerConsumer,
    dailyCallsTotal: descriptorToCompare.dailyCallsTotal,
  }

  return getKeys(newDescriptorData).every((key) =>
    isEqual(newDescriptorData[key], descriptorDataToCompare[key])
  )
}

export function compareAttributesStep3IfEquals(
  newAttributes: RemappedEServiceAttributes,
  attributesToCompare: DescriptorAttributes
) {
  const remappedAttributesToCompare = remapEServiceAttributes(attributesToCompare)

  const newAttributesToCompare = {
    certified: newAttributes.certified.filter((group) => group.attributes.length > 0),
    verified: newAttributes.verified.filter((group) => group.attributes.length > 0),
    declared: newAttributes.declared.filter((group) => group.attributes.length > 0),
  }

  return getKeys(newAttributes).every((key) =>
    isEqual(newAttributesToCompare[key], remappedAttributesToCompare[key])
  )
}
