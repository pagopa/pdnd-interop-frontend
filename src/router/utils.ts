import { routes } from '@/router/routes'
import type { LangCode, ProviderOrConsumer } from '@/types/common.types'
import type { RouteConfig, RouteKey } from '@/router/types'
import { generatePath, matchPath } from 'react-router-dom'
import { getKeys } from '@/utils/array.utils'
import memoize from 'lodash/memoize'
import identity from 'lodash/identity'
import sortBy from 'lodash/sortBy'
import { PUBLIC_URL } from '@/config/env'
import qs from 'qs'
import QueryString from 'qs'

/** Returns the localized path of the given routeKey and language  */
export function getLocalizedRoutePathname(routeKey: RouteKey, lang: LangCode) {
  return `${PUBLIC_URL}/${lang}/${routes[routeKey].PATH[lang]}`
}

/** Checks if the routeKey represent the given pathname */
function matchRouteKeyPath(pathname: string, lang: LangCode, routeKey: RouteKey) {
  return matchPath(getLocalizedRoutePathname(routeKey as RouteKey, lang), pathname)
}

/** Returns the routeKey of the given pathname */
export const getRouteKeyFromPathname = memoize(function _getRouteKeyFromPathname(
  pathname: string,
  lang: LangCode
) {
  let pathnameWithBaseUrl = pathname
  if (!pathnameWithBaseUrl.startsWith(PUBLIC_URL)) {
    pathnameWithBaseUrl = PUBLIC_URL + pathnameWithBaseUrl
  }
  const isTheRouteKeyPath = matchRouteKeyPath.bind(null, pathnameWithBaseUrl, lang)
  const routeKey = getKeys(routes).find(isTheRouteKeyPath)

  if (!routeKey) {
    throw new Error(`Pathname ${pathnameWithBaseUrl} has no associated routeKey.`)
  }

  return routeKey
})

export function switchPathLang(fromLang: LangCode, toLang: LangCode) {
  for (const routeKey of Object.keys(routes)) {
    const match = matchRouteKeyPath(window.location.pathname, fromLang, routeKey as RouteKey)

    if (match) {
      let newPath = generatePath(
        getLocalizedRoutePathname(routeKey as RouteKey, toLang),
        match.params as Record<string, string>
      )

      if (window.location.search) {
        newPath += window.location.search
      }

      if (window.location.hash) {
        newPath += window.location.hash
      }

      window.location.replace(newPath)
      return
    }
  }
}

export const URL_FRAGMENTS: Record<string, Record<LangCode, string>> = {
  FIRST_DRAFT: { it: 'prima-bozza', en: 'first-draft' },
  EDIT: { it: 'modifica', en: 'bozza' },
}

export function getSplittedPath(route: RouteConfig, currentLang: LangCode) {
  return route.PATH[currentLang].split('/').filter(identity)
}

export const getParentRoutes = memoize((routeKey: RouteKey): Array<RouteKey> => {
  function isParentRoute(
    possibleParentRouteSubpaths: Array<string>,
    currentSubpaths: Array<string>
  ) {
    if (possibleParentRouteSubpaths.length >= currentSubpaths.length) {
      return false
    }

    const allSameFragments = possibleParentRouteSubpaths.every(
      (pathFragment, i) => pathFragment === currentSubpaths[i]
    )

    // URL_FRAGMENTS.EDIT is appended at the end of a read path of the same type.
    // E.g. /eservice/myid becomes /eservice/myid/edit. So it's always same length + 1
    const isFalseParent = currentSubpaths.length === possibleParentRouteSubpaths.length + 1

    if (allSameFragments && isEditPath(routeKey) && isFalseParent) {
      return false
    }

    return allSameFragments
  }

  const route = routes[routeKey]
  const currentSubpaths = getSplittedPath(route, 'it')

  const parents = Object.entries(routes).filter(([_, possibleParentRoute]) =>
    isParentRoute(getSplittedPath(possibleParentRoute, 'it'), currentSubpaths)
  )

  const sortedParents = sortBy(parents, ([_, parent]) => getSplittedPath(parent, 'it').length)

  return sortedParents.map(([routeKey]) => routeKey) as Array<RouteKey>
})

export const isProviderOrConsumerRoute = memoize(
  (routeKey: RouteKey): ProviderOrConsumer | null => {
    const excludeList = ['ui', 'it']
    const subroutes = getSplittedPath(routes[routeKey], 'it').filter(
      (b) => !excludeList.includes(b)
    )
    const [mode] = subroutes

    if (mode === 'erogazione') {
      return 'provider'
    }

    if (mode === 'fruizione') {
      return 'consumer'
    }

    return null
  }
)

export const isEditPath = memoize((routeKey: RouteKey): boolean => {
  const subroutes = getSplittedPath(routes[routeKey], 'it')

  const lastBit = subroutes[subroutes.length - 1]
  return Object.values(URL_FRAGMENTS.EDIT).some((f) => lastBit.endsWith(f))
})

export function parseSearch(search: string) {
  return qs.parse(search, { ignoreQueryPrefix: true })
}

export function stringifySearch(searchObj: Record<string, string> | QueryString.ParsedQs) {
  return qs.stringify(searchObj)
}

/**
 * Returns an array with all the dynamic path names for a given RouteKey
 * @example
 * "/:foo/test/:bar" => ["foo", "bar"]
 */
export const getDynamicPathsName = memoize((routeKey: RouteKey) => {
  return routes[routeKey].PATH.it
    .split('/')
    .filter(identity)
    .filter((subpath) => subpath.startsWith(':'))
    .map((param) => param.replace(':', ''))
})
