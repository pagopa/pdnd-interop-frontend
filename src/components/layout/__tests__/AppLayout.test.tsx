import React from 'react'
import { AppLayout } from '../AppLayout'
import { renderWithApplicationContext } from '@/utils/testing.utils'
import { vi } from 'vitest'

vi.mock('../Breadcrumbs', () => {
  return { Breadcrumbs: () => <></> }
})

describe('AppLayout', () => {
  it('should match snapshot', () => {
    const screen = renderWithApplicationContext(<AppLayout>{}</AppLayout>, {
      withRouterContext: true,
    })
    expect(screen.baseElement).toMatchSnapshot()
  })

  it('should match snapshot (hideSideNav)', () => {
    const screen = renderWithApplicationContext(<AppLayout hideSideNav>{}</AppLayout>, {
      withRouterContext: true,
    })
    expect(screen.baseElement).toMatchSnapshot()
  })
})
