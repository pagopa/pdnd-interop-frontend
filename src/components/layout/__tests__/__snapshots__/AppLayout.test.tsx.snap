// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppLayout > should match snapshot (hideSideNav) 1`] = `
<body>
  <div>
    <main
      class="MuiBox-root css-ktdqii"
    >
      <div
        class="MuiBox-root css-u24siw"
      />
    </main>
  </div>
</body>
`;

exports[`AppLayout > should match snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-1y6z04t"
    >
      <div
        class="css-1e5vhfl-MuiStack-root"
      >
        <nav
          class="MuiBox-root css-15sxj0r"
        >
          <ul
            class="MuiList-root css-1dc7nhp-MuiList-root"
          >
            <div
              class="MuiBox-root css-0"
            >
              <li
                class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-15zi390-MuiListItem-root"
              >
                <div
                  class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                  >
                    <span
                      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
                      style="width: 120px;"
                    />
                  </p>
                </div>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-colorDisabled MuiSvgIcon-fontSizeMedium css-1db085k-MuiSvgIcon-root"
                  data-testid="ExpandMoreIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
              </li>
              <ul
                class="MuiList-root css-13i2vi2-MuiList-root"
              />
            </div>
            <div
              class="MuiBox-root css-0"
            >
              <li
                class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-15zi390-MuiListItem-root"
              >
                <div
                  class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                  >
                    <span
                      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
                      style="width: 120px;"
                    />
                  </p>
                </div>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-colorDisabled MuiSvgIcon-fontSizeMedium css-1db085k-MuiSvgIcon-root"
                  data-testid="ExpandLessIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m12 8-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z"
                  />
                </svg>
              </li>
              <ul
                class="MuiList-root css-13i2vi2-MuiList-root"
              >
                <li
                  class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1nlpcw4-MuiListItem-root"
                >
                  <div
                    class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      <span
                        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
                        style="width: 160px;"
                      />
                    </p>
                  </div>
                </li>
                <li
                  class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1nlpcw4-MuiListItem-root"
                >
                  <div
                    class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      <span
                        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
                        style="width: 160px;"
                      />
                    </p>
                  </div>
                </li>
                <li
                  class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1nlpcw4-MuiListItem-root"
                >
                  <div
                    class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      <span
                        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
                        style="width: 160px;"
                      />
                    </p>
                  </div>
                </li>
                <li
                  class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1nlpcw4-MuiListItem-root"
                >
                  <div
                    class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      <span
                        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
                        style="width: 160px;"
                      />
                    </p>
                  </div>
                </li>
                <li
                  class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1nlpcw4-MuiListItem-root"
                >
                  <div
                    class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      <span
                        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
                        style="width: 160px;"
                      />
                    </p>
                  </div>
                </li>
              </ul>
            </div>
            <li
              class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1nlpcw4-MuiListItem-root"
            >
              <div
                class="MuiListItemText-root css-tlelie-MuiListItemText-root"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
                    style="width: 160px;"
                  />
                </p>
              </div>
            </li>
            <li
              class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1nlpcw4-MuiListItem-root"
            >
              <div
                class="MuiListItemText-root css-tlelie-MuiListItemText-root"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
                    style="width: 160px;"
                  />
                </p>
              </div>
            </li>
          </ul>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-1b008ma-MuiDivider-root"
          />
          <ul
            class="MuiList-root MuiList-padding css-1y8vw3r-MuiList-root"
          >
            <li
              class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1nlpcw4-MuiListItem-root"
            >
              <div
                class="MuiListItemText-root css-tlelie-MuiListItemText-root"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
                    style="width: 160px;"
                  />
                </p>
              </div>
            </li>
          </ul>
        </nav>
        <div
          class="MuiBox-root css-ebpwpv"
        >
          <main
            class="MuiBox-root css-1jl3o0g"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
